Nakama/PresenceUtil.cs

```diff
@@ -0,0 +1,74 @@
+// Copyright 2023 The Nakama Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Collections.Generic;
+
+namespace Nakama
+{
+    internal static class PresenceUtil
+    {
+        /// <summary>
+        /// Applies joins and leaves of a presence event to a copy of the provided presence list and returns the result.
+        /// </summary>
+        public static List<UserPresence> CopyJoinsAndLeaves(List<UserPresence> currentPresences, IEnumerable<IUserPresence> joins, IEnumerable<IUserPresence> leaves)
+        {
+            currentPresences = currentPresences ?? new List<UserPresence>();
+            joins = joins ?? new List<UserPresence>();
+            leaves = leaves ?? new List<UserPresence>();
+
+            var newPresences = new Dictionary<string, UserPresence>();
+
+            foreach (UserPresence presence in currentPresences)
+            {
+                newPresences[presence.UserId] = presence;
+            }
+
+            foreach (IUserPresence join in joins)
+            {
+                if (newPresences.ContainsKey(join.UserId))
+                {
+                    // unexpected
+                    continue;
+                }
+
+                newPresences.Add(join.UserId, IUserPresenceToUserPresence(join));
```

Comment:
Why is `IUserPresenceToUserPresence` helper method necessary? Can we not just cast the join presence?

```suggestion
                newPresences.Add(join.UserId, (UserPresence) join);
```

Reply:
That would work for now because there's a 1:1 mapping between our DTOs and interfaces but if we add a second type of DTO (e.g., for protobuf) that would break.

Reply:
It would work in most cases except if the user passed in their own implementation of `IUserPresence.`

In reality that will almost never happen but it was an easy enough case to cover.

Reply:
@lugehorsam What do you think about a fast path that uses `is` to check if we can cast to the type we know it will be in most cases and if not fall back on the helper function?

Reply:
sure 👍